File: src/jvm/main/org/jetbrains/kotlinx/lincheck/paramgen/StringGen.java
/*
 * Lincheck
 *
 * Copyright (C) 2019 - 2023 JetBrains s.r.o.
 *
 * This Source Code Form is subject to the terms of the
 * Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
 * with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package org.jetbrains.kotlinx.lincheck.paramgen;

import org.jetbrains.kotlinx.lincheck.RandomProvider;

import java.util.Random;

public class StringGen implements ParameterGenerator<String> {
    private static final int DEFAULT_MAX_WORD_LENGTH = 15;
    private static final String DEFAULT_ALPHABET = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_ ";

    private final Random random;
    private final int maxWordLength;
    private final String alphabet;
    private int currentWordLength = 1;

    public StringGen(RandomProvider randomProvider, String configuration) {
        random = randomProvider.createRandom();
        if (configuration.isEmpty()) { // use default configuration
            maxWordLength = DEFAULT_MAX_WORD_LENGTH;
            alphabet = DEFAULT_ALPHABET;
            return;
        }
        int firstCommaIndex = configuration.indexOf(':');
        if (firstCommaIndex < 0) { // maxWordLength only
            maxWordLength = Integer.parseInt(configuration);
            alphabet = DEFAULT_ALPHABET;
        } else { // maxWordLength:alphabet
            maxWordLength = Integer.parseInt(configuration.substring(0, firstCommaIndex));
            alphabet = configuration.substring(firstCommaIndex + 1);
        }
    }

    public String generate() {
        if (currentWordLength < maxWordLength && random.nextBoolean()) {
            currentWordLength++;
        }

        char[] cs = new char[currentWordLength];
        for (int i = 0; i < cs.length; i++) {
            cs[i] = alphabet.charAt(random.nextInt(alphabet.length()));
        }

        return new String(cs);
    }

    @Override
    public void resetRange() {
        currentWordLength = 1;
    }
}

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/paramgen/StringGen.kt
/*
* #%L
* Lincheck
* %%
* Copyright (C) 2015 - 2018 Devexperts, LLC
* %%
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License as
* published by the Free Software Foundation, either version 3 of the
* License, or (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Lesser Public License for more details.
*
* You should have received a copy of the GNU General Lesser Public
* License along with this program.  If not, see
* <http://www.gnu.org/licenses/lgpl-3.0.html>.
* #L%
*/
package org.jetbrains.kotlinx.lincheck.paramgen

import java.util.*

class StringGen(configuration: String) : ParameterGenerator<String?> {
    private val random = Random(0)
    private var maxWordLength = 0
    private var alphabet: String? = null
    override fun generate(): String {
        val cs = CharArray(random.nextInt(maxWordLength))
        for (i in cs.indices) cs[i] = alphabet!![random.nextInt(alphabet!!.length)]
        return String(cs)
    }

    companion object {
        private const val DEFAULT_MAX_WORD_LENGTH = 15
        private const val DEFAULT_ALPHABET = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_ "
    }

    init {
        if (configuration.isEmpty()) { // use default configuration
            maxWordLength = DEFAULT_MAX_WORD_LENGTH
            alphabet = DEFAULT_ALPHABET
        } else {
            val firstCommaIndex = configuration.indexOf(':')
            if (firstCommaIndex < 0) { // maxWordLength only
                maxWordLength = configuration.toInt()
                alphabet = DEFAULT_ALPHABET
            } else { // maxWordLength:alphabet
                maxWordLength = configuration.substring(0, firstCommaIndex).toInt()
                alphabet = configuration.substring(firstCommaIndex + 1)
            }
        }
    }
}